MG-2380 - Pass env files to vault script

JeffMboya opened the pull request:
<!-- Copyright (c) Abstract Machines
SPDX-License-Identifier: Apache-2.0 -->

<!--

Pull request title should be `MG-XXX - description` or `NOISSUE - description` where XXX is ID of the issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/absmach/magistrala/blob/main/CONTRIBUTING.md) file for detailed contributing guidelines.

For Work In Progress Pull Requests, please use the Draft PR feature, see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments.

- Provide tests for your changes.
- Use descriptive commit messages.
- Comment your code where appropriate.
- Squash your commits
- Update any related documentation.
-->

# What type of PR is this?

<!--This represents the type of PR you are submitting.

For example:
This is a bug fix because it fixes the following issue: #1234
This is a feature because it adds the following functionality: ...
This is a refactor because it changes the following functionality: ...
This is a documentation update because it updates the following documentation: ...
This is a dependency update because it updates the following dependencies: ...
This is an optimization because it improves the following functionality: ...
-->

This is a feature because it adds the functionality to pass the `.env` file path as CLI arguments to Vault scripts.

## What does this do?

<!--
Please provide a brief description of what this PR is intended to do.
Include List any changes that modify/break current functionality.
-->

This PR modifies the Vault scripts to support passing the `.env` file path as a command-line argument. It updates the `readDotEnv` function to use the provided `.env` file path and handles the default path if no argument is provided.

## Which issue(s) does this PR fix/relate to?

<!--
For pull requests that relate or close an issue, please include them below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "Resolves #1234" would connect the current pull request to issue 1234.  And when we merge the pull request, Github will automatically close the issue.
-->

- Related Issue https://github.com/absmach/devops/issues/135
- Resolves #2380

## Have you included tests for your changes?

<!--If you have not included tests, please explain why.
For example:
No
-->

No
## Did you document any new/modified feature?

<!--If you have not included documentation, please explain why.
For example:
Yes, I have updated the documentation for the new feature.
No, I have not updated the documentation because I do not know how to.
-->

Yes, I have updated the documentation for the new feature.

### Notes

<!--Please provide any additional information you feel is important.-->

This update ensures that the Vault scripts are more flexible and can work with `.env` files located in different paths.

JeffMboya committed:
chore: Update .env file path in vault scripts

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update documentation to include custom  file path support for Vault scripts

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update README

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

chore: Update README

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

move Vault scripts to docker/addons/vault/scripts

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Replace MAGISTRALA_DIR with relative paths

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update scripts to use docker/addons/vault/scripts

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update .env file relative path in Vault scripts

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update .env file relative path in Vault scripts

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update .env file relative path in Vault scripts

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Remove scripts directory

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

arvindh123 commented on docker/addons/vault/vault_copy_certs.sh:
```diff
@@ -9,9 +9,21 @@ export MAGISTRALA_DIR=$scriptdir/../../../
+    shift
+done
+
 readDotEnv() {
```
So something like this we can try 
```suggestion

# Default .env file path
env_file="docker/.env"

# Parse command line arguments
while [[ "$#" -gt 0 ]]; do
    case $1 in
        --env-file) env_file="$2"; shift ;;
        *) echo "Unknown parameter passed: $1"; exit 1 ;;
    esac
    shift
done

certs_copy_path="docker/ssl/certs/"
while [[ "$#" -gt 0 ]]; do
    case $1 in
        --dest-path) certs_copy_path="$2"; shift ;;
        *) echo "Unknown parameter passed: $1"; exit 1 ;;
    esac
    shift
done


readDotEnv() {
    set -o allexport
    source "$env_file"
    set +o allexport
}

readDotEnv

server_name="localhost"

# Check if MG_NGINX_SERVER_NAME is set or not empty
if [ -n "${MG_NGINX_SERVER_NAME:-}" ]; then
    server_name="$MG_NGINX_SERVER_NAME"
fi

echo "Copying certificate files"

if [ -e "data/${server_name}.crt" ]; then
    cp -v data/${server_name}.crt      ${certs_copy_path}/magistrala-server.crt
else
    echo "${server_name}.crt file not available"
fi

if [ -e "data/${server_name}.key" ]; then
    cp -v data/${server_name}.key      ${certs_copy_path}/magistrala-server.key
else
    echo "${server_name}.key file not available"
fi

if [ -e "data/${MG_VAULT_PKI_INT_FILE_NAME}.key" ]; then
    cp -v data/${MG_VAULT_PKI_INT_FILE_NAME}.key    ${certs_copy_path}/ca.key
else
    echo "data/${MG_VAULT_PKI_INT_FILE_NAME}.key file not available"
fi

if [ -e "data/${MG_VAULT_PKI_INT_FILE_NAME}_bundle.crt" ]; then
    cp -v data/${MG_VAULT_PKI_INT_FILE_NAME}_bundle.crt     ${certs_copy_path}/ca.crt
else
    echo "data/${MG_VAULT_PKI_INT_FILE_NAME}_bundle.crt file not available"
fi

exit 0

```

arvindh123 replied:
[> So something like this we can try](https://github.com/absmach/magistrala/pull/2387#discussion_r1726947534)

After implementing this please  check it's working  with https://github.com/absmach/amdm/pull/88

arvindh123 replied:
@JeffMboya Please move all the scripts to a directory

JeffMboya replied:
@JeffMboya 
@arvindh123 A quick clarification: move the scripts to a single directory inside the vault folder i.e `docker/addons/vault/scripts` or outside the vault folder i.e magistrala/scripts? https://github.com/absmach/magistrala/pull/2387#discussion_r1727039272


@Arvindh
Good question,
I think we can move to `docker/addons/vault/scripts` for now
What do you think?

@JeffMboya
I'm thinking of using `docker/addons/vault/scripts` because it keeps the scripts together with other Vault files.

@arvindh123
okay then lets go with `docker/addons/vault/scripts`

JeffMboya replied:
All the vault scripts have been moved to `docker/addons/vault/scripts`

JeffMboya committed:
Move scripts into a single directory

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

JeffMboya committed:
Update README

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update README

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Add empty line

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Remove empty line

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

arvindh123 commented on docker/addons/vault/scripts/vault_init.sh:
```diff
@@ -0,0 +1,36 @@
+# Create the data directory inside docker/addons/vault/scripts
+mkdir -p data
+
+vault operator init -address=$MG_VAULT_ADDR 2>&1 | tee >(sed -r 's/\x1b\[[0-9;]*m//g' > data/secrets)
```
Add new line

arvindh123 commented on docker/addons/vault/docker-compose.yml:
```diff
@@ -26,7 +26,7 @@ services:
       - magistrala-vault-volume:/vault/logs
       - ./config.hcl:/vault/config/config.hcl
-      - ./entrypoint.sh:/entrypoint.sh
+      - ./scripts/entrypoint.sh:/entrypoint.sh
```
@JeffMboya This can be in `docker/addons/vault`, becuase it is script for docker , and not for setup

arvindh123 reviewed (changes requested):
Move docker/addons/vault/.gitignore to scripts directory

arvindh123 commented on docker/addons/vault/scripts/vault_copy_certs.sh:
```diff
@@ -5,13 +5,27 @@
+        *) echo "Unknown parameter passed: $1"; exit 1 ;;
+    esac
+    shift
+done
```
Please add options to pass `certs_copy_path`

JeffMboya committed:
Updated the script to accept --certs-copy-path command-line argument

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

arvindh123 reviewed (changes requested):
For this bash command from root of repo 
```bash 
docker/addons/vault/scripts/vault_init.sh --env-file docker/.env
````
I get below error
```bash
docker/addons/vault/scripts/vault_init.sh: line 25: docker/.env: No such file or directory
```

JeffMboya committed:
Update .env file relative path in Vault scripts

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

JeffMboya committed:
Update Vault environment variables in vault_copy_env.sh script

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update Vault environment variables in vault_copy_env.sh script

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

JeffMboya committed:
Update Vault environment variables in .env file

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update Vault environment variables in vault_init.sh

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update Vault environment variables in vault_init.sh

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update Vault environment variables in vault_init.sh

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update Vault environment variables in vault_copy_env.sh

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Add empty line

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update env path in vault_unseal.sh

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update env path

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update Vault environment variables in .env file

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

arvindh123 commented on docker/addons/vault/scripts/vault_init.sh:
```diff
@@ -0,0 +1,41 @@
+
+scriptdir="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+repo_root="$(realpath "$scriptdir/../../../../")"
+env_file="$repo_root/docker/.env"
```
Let the `env_file` path be relative `docker/.env` and remove the `repo_root`

arvindh123 commented on docker/addons/vault/scripts/vault_set_pki.sh:
```diff
@@ -58,9 +77,9 @@ vaultGenerateRootCACertificate() {
-        ttl=87600h | tee >(jq -r .data.certificate >data/${MG_VAULT_PKI_FILE_NAME}_ca.crt) \
-                         >(jq -r .data.issuing_ca  >data/${MG_VAULT_PKI_FILE_NAME}_issuing_ca.crt) \
-                         >(jq -r .data.private_key >data/${MG_VAULT_PKI_FILE_NAME}_ca.key)
+        ttl=87600h | tee >(jq -r .data.certificate >"$repo_root/docker/addons/vault/scripts/data/${MG_VAULT_PKI_FILE_NAME}_ca.crt") \
```
Instead of `$repo_root/docker/addons/vault/scripts/data` use `$scriptdir/data`
And remove `$repo_root`

arvindh123 commented on docker/addons/vault/scripts/vault_copy_env.sh:
```diff
@@ -0,0 +1,42 @@
+while [[ "$#" -gt 0 ]]; do
+    case $1 in
+        --env-file)
+            if [[ "$2" = /* ]]; then
```
@JeffMboya What does it does ? 
```bash
if [[ "$2" = /* ]]; then
```
If i pass /* to --env-file 
```bash
docker/addons/vault/scripts/vault_copy_env.sh --env-file /*
```
then getting a wired error 
```
Unknown parameter passed: /bin
```

arvindh123 replied:
Simply we can throw error , If `$2` is empty.

JeffMboya replied:
@arvindh123 This checks if the second argument (`$2`) is an absolute path, which in Unix/Linux means it starts with a `/`. If it is an absolute path, the script simply assigns it to the `env_file` variable as it is.

JeffMboya replied:
The error happens because `/*` is a wildcard pattern that matches all directories and files at the root level such as `/bin`, `/etc`, `/lib`, etc., and the script is picking up `/bin` as one of those matches. The script then attempts to process `/bin` as the value of `--env-file`, which it doesn't expect, leading to the `"Unknown parameter passed: /bin"` error.

arvindh123 replied:
Why do we need this logic ? Just trying to understand you code 
Particularly why we need to convert relative path to Absolute Path ? 
```bash
            if [[ "$2" = /* ]]; then
                env_file="$2"
            else
                env_file="$(realpath -m "$PWD/$2")"
            fi
            shift
            ;;
```

JeffMboya replied:
> Simply we can throw error , If `$2` is empty.

@arvindh123 an error is thrown when `$2` is empty:
```sh
$ magistrala/docker/addons/vault/scripts/vault_copy_env.sh --env-file   
magistrala/docker/addons/vault/scripts/vault_copy_env.sh: line 15: $2: unbound variable
```

JeffMboya replied:
> Why do we need this logic ? Just trying to understand you code Particularly why we need to convert relative path to Absolute Path ?
> 
> ```shell
>             if [[ "$2" = /* ]]; then
>                 env_file="$2"
>             else
>                 env_file="$(realpath -m "$PWD/$2")"
>             fi
>             shift
>             ;;
> ```

The logic is included in the script to ensure that the path to the environment file is always fully qualified and unambiguous.

arvindh123 replied:
Even it was fully qualified , without check it existence , we could not say that file is exists . 
Even it is relative path , if file exists is checked ,then we need to verify it is text file of env kind. (but we can skip this verification but file existence is needed)

JeffMboya replied:
> Even it was fully qualified , without check it existence , we could not say that file is exists .
> Even it is relative path , if file exists is checked ,then we need to verify it is text file of env kind. (but we can skip this verification but file existence is needed)

Yep, that makes sense. The most important part is making sure the file is there. I think the extra check is nice to have but not absolutely necessary.

JeffMboya committed:
Add error for empty --env-file and --certs-copy-path values

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Add a new line

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

arvindh123 commented on docker/addons/vault/scripts/vault_copy_certs.sh:
```diff
@@ -0,0 +1,86 @@
+            fi
+            shift
+            ;;
+        --certs-copy-path)
```
For the below command i get error
```bash
docker/addons/vault/scripts/vault_copy_env.sh --env-file docker/.env --certs-copy-path ~/mainflux/
```

Error Message
```log 
Unknown parameter passed: --certs-copy-path
```

JeffMboya replied:
@arvindh123 I am not able to recreate this error

``` 
$docker/addons/vault/scripts/vault_copy_certs.sh --env-file docker/.env --certs-copy-path certs/pki     
Copying certificate files to certs/pki
'/home/jeff/Documents/magistrala/docker/addons/vault/scripts/data/localhost.crt' -> 'certs/pkimagistrala-server.crt'
'/home/jeff/Documents/magistrala/docker/addons/vault/scripts/data/localhost.key' -> 'certs/pkimagistrala-server.key'
'/home/jeff/Documents/magistrala/docker/addons/vault/scripts/data/mg_int.key' -> 'certs/pkica.key'
'/home/jeff/Documents/magistrala/docker/addons/vault/scripts/data/mg_int_bundle.crt' -> 'certs/pkica.crt'
```

JeffMboya replied:
@arvindh123 `vault_copy_env.sh` only accepts `--env-file` argument
To pass both `--env-file` and `--certs-copy-path` argument, run `vault_copy_certs.sh`

arvindh123 commented on docker/addons/vault/scripts/vault_create_approle.sh:
```diff
@@ -4,39 +4,62 @@
+            if [[ ! -f "$env_file" ]]; then
+                echo "Error: .env file not found at $env_file"
+                exit 1
+            fi
```
Please add  flage`--skip-enable-app-role` to set  `SKIP_ENABLE_APP_ROLE`

arvindh123 commented on docker/addons/vault/scripts/vault_copy_certs.sh:
```diff
@@ -0,0 +1,86 @@
+    echo "$scriptdir/data/${MG_VAULT_PKI_INT_FILE_NAME}_bundle.crt file not available"
+fi
+
+exit 0
```
add new line here

JeffMboya committed:
Update SKIP_ENABLE_APP_ROLE initialization in vault_create_approle.sh

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

Update SKIP_ENABLE_APP_ROLE initialization in vault_create_approle.sh

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>

dborovcanin commented on docker/addons/vault/README.md:
```diff
@@ -45,22 +45,22 @@ When the Vault service is started, some initialization steps need to be done to
+Unseal Key 2: `<UNSEAL_KEY_2>`
+Unseal Key 3: `<UNSEAL_KEY_3>`
+Unseal Key 4: `<UNSEAL_KEY_4>`
+Unseal Key 5: `<UNSEAL_KEY_5>`
```
Revert this change. While placeholders are self-explanatory, this conflicts with the `example` premise (i.e. this is not an example anymore and tells nothing about how keys look like).

JeffMboya committed:
Update README

Signed-off-by: JeffMboya <jangina.mboya@gmail.com>
